connection.go

```diff
@@ -547,6 +570,155 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
 	return respd, nil
 }
 
+func isAsyncMode(ctx context.Context) (bool, error) {
+	val := ctx.Value(AsyncMode)
+	if val == nil {
+		return false, nil
+	}
+	boolVal, ok := val.(bool)
+	if !ok {
+		return false, fmt.Errorf("failed to cast val %+v to bool", val)
+	}
+	return boolVal, nil
+}
+
+func getAsync(
+	ctx context.Context,
+	sr *snowflakeRestful,
+	headers map[string]string,
+	URL *url.URL,
+	timeout time.Duration,
+	res *snowflakeResult,
+	rows *snowflakeRows,
+) {
+	resType, _ := ctx.Value(resultType).(string)
+	headers[headerAuthorizationKey] = fmt.Sprintf(headerSnowflakeToken, sr.Token)
+	resp, err := sr.FuncGet(ctx, sr, URL, headers, timeout)
+	if err != nil {
+		logger.WithContext(ctx).Errorf("failed to get response. err: %v", err)
+		if resType == "exec" {
+			res.statusChannel <- queryEvent{
+				QueryFailed,
+				&SnowflakeError{
+					Number:  -1,
+					Message: err.Error(),
+					QueryID: res.queryID,
+				},
+			}
+			close(res.statusChannel)
+		} else {
+			rows.statusChannel <- queryEvent{
+				QueryFailed,
+				&SnowflakeError{
+					Number:  -1,
+					Message: err.Error(),
+					QueryID: rows.queryID,
+				},
+			}
+			close(rows.statusChannel)
+		}
+		return
+	}
+
+	respd := execResponse{}
+	err = json.NewDecoder(resp.Body).Decode(&respd)
+	resp.Body.Close()
```

Comment:
There's a small risk of a memory leak here, but I think it can be easily mitigated this way after calling FuncGet() above:

```
if resp.Body != nil {
  defer resp.Body.Close()
}
```

I _think_ (please verify) that it's okay to call resp.Body.Close() multiple times so you can still leave this close call in here.

The memory leak risk is there's a lot of code between this close call and the response body being created. If anything panics along the way (recovered later by the user) or if FuncGet() doesn't guarantee that `err != nil && resp = nil` -- you might have a leak that's hard to track down.

Reply:
I confirmed that it's okay to call `.Close()` on a ReadCloser.